src/solver/simulation/hydro-final-reservoir-level-pre-checks.cpp

```diff
@@ -0,0 +1,115 @@
+    for (uint tsIndex = 0; tsIndex != study.scenarioFinalHydroLevels.height; ++tsIndex)
+    {
+        study.areas.each(
+          [&](Data::Area& area)
+          {
+              double deltaReservoirLevel = 0.0;
+              // TODO CR25:
+              /*at this point the pre-checks are done for all MC before running the simulation
+              and the simulation is ended immediately not waisting user time!
+              However, at this point "tsIndex = (uint)ptchro.Hydraulique" is not assigned yet
+              so the pre-checks are done for all MC years (whether they are used or not in hydro
+              scenario-builder) this can lead to error reporting for MC years that are not used (not
+              the case when pre-checks done im management.cpp)
+              */
+              // auto& ptchro = *NumeroChroniquesTireesParPays[numSpace][area.index];
+              auto& inflowsmatrix = area.hydro.series->storage;
+              // auto tsIndex = (uint)ptchro.Hydraulique;
+              auto const& srcinflows = inflowsmatrix[tsIndex < inflowsmatrix.width ? tsIndex : 0];
+              double initialReservoirLevel = study.scenarioHydroLevels[area.index][tsIndex];
+              double finalReservoirLevel = study.scenarioFinalHydroLevels[area.index][tsIndex];
+
+              if (area.hydro.reservoirManagement && !area.hydro.useWaterValue
+                  && !isnan(finalReservoirLevel) && !isnan(initialReservoirLevel))
+              {
+                  // deltaReservoirLevel
+                  deltaReservoirLevel = initialReservoirLevel - finalReservoirLevel;
+                  // collect data for pre-checks
+                  uint initReservoirLvlMonth
+                    = area.hydro.initializeReservoirLevelDate; // month [0-11]
+                  uint initReservoirLvlDay
+                    = study.calendar.months[initReservoirLvlMonth].daysYear.first;
+                  double reservoirCapacity = area.hydro.reservoirCapacity;
+                  double lowLevelLastDay
+                    = area.hydro.reservoirLevel[Data::PartHydro::minimum][DAYS_PER_YEAR - 1];
+                  double highLevelLastDay
+                    = area.hydro.reservoirLevel[Data::PartHydro::maximum][DAYS_PER_YEAR - 1];
+                  double totalYearInflows = 0.0;
+                  // calculate yearly inflows
+                  for (uint day = initReservoirLvlDay; day < DAYS_PER_YEAR; ++day)
+                  {
+                      totalYearInflows += srcinflows[day];
+                  }
+                  // pre-check 1 -> reservoir_levelDay_365 – reservoir_levelDay_1 ≤
+                  // yearly_inflows
+                  if ((finalReservoirLevel - initialReservoirLevel) * reservoirCapacity
+                      > totalYearInflows) // ROR time-series in MW (power), SP time-series in MWh (energy)
+                  {
+                      logs.error() << "Year: " << tsIndex + 1 << ". Area: " << area.name
```

Comment:
tsIndex is not a year index, it's a time-series index. The mapping year -> tsIndex is given by the sc. builder.